addons/website_sale/controllers/main.py

```diff
@@ -234,10 +234,36 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
 
         Category = request.env['product.public.category']
         search_categories = False
-        search_product = Product.search(domain)
         if search:
-            categories = search_product.mapped('public_categ_ids')
-            search_categories = Category.search([('id', 'parent_of', categories.ids)] + request.website.website_domain())
+            # subquery for search_product = Product.search(domain)
+            query = Product._where_calc(domain)
+            Product._apply_ir_rules(query, 'read')
+            prod_from, prod_where, prod_params = query.get_sql()
+            prod_query = 'SELECT "{}".id FROM {} WHERE {}'.format(Product._table, prod_from, prod_where or "true")
+
+            # categories = search_product.mapped('public_categ_ids')
+            Category.check_access_rights('read')
+            query = Category._where_calc([])
+            Category._apply_ir_rules(query, 'read')
+            cat_from, cat_where, cat_params = query.get_sql()
+
+            field = type(Product).public_categ_ids
+
+            query = """
+                SELECT "{cat_table}".id FROM {cat_from}
+                WHERE {cat_where} AND "{cat_table}".id IN (
+                    SELECT rel.{rel_cat} FROM {rel_table} rel
+                    WHERE rel.{rel_prod} IN ({prod_query})
```

Comment:
Adding a SELECT in a WHERE of a SELECT that is in a WHERE is possibly the worst way to optimize things. Try doing it with JOINs wherever possible.